docs/design/simple-rolling-update.md

```diff
@@ -0,0 +1,84 @@
+## Simple rolling update
+This is a lightweight design document for simple rolling update in ```kubectl```
+
+### Lightweight rollout
+Assume that we have a current replication controller named ```foo``` and it is running image ```image:v1```
+
+```kubectl rolling-update rc foo [foo-v2] --image=myimage:v2```
+
+If the user doesn't specify a name for the 'next' controller, then the 'next' controller is renamed to
+the name of the original controller.
+
+Obviously there is a race here, where if you kill the client between delete foo, and creating the new version of 'foo' you might be surprised about what is there, but I think that's ok.
+See Recovery below
+
+If the user does specify a name for the 'next' controller, then the 'next' controller is retained with its existing name,
+and the old 'foo' controller is deleted.
+
+#### Recovery
+If a rollout fails or is terminated in the middle, it is important that the user be able to resume the roll out.
+To facilitate recovery in the case of a crash of the updating process itself, we add the following annotations to each replicaController:
+   * ```desired-replicas``` The desired number of replicas for this controller (either N or zero)
+   * ```update-partner``` A pointer to the resource that is the other half of this update (syntax ```<namespace>:<id>```)
+
+Recovery is achieved by issuing the same command again:
+
+```
+kubectl rolling-update rc foo [foo-v2] --image=myimage:v2
+```
+
+Whenever the rolling update command executes, the kubectl client looks for replication controllers called ```foo``` and ```foo-next```, if the exist, an attempt is
+made to roll ```foo``` to ```foo-next```.  If ```foo-next``` does not exist, then it is created, and the rollout is a new rollout.  If ```foo``` doesn't exist, then
+it is assumed that the rollout is nearly completed, and ```foo-next``` is renamed to ```foo```.  Details of the execution flow are given below.
+
+
+### Aborting a rollout
+Abort is assumed to want to reverse a rollout in progress.
+
+```kubectl rolling-update rc foo [foo-v2] abort```
```

Comment:
> On Apr 22, 2015, at 1:17 PM, Brendan Burns <notifications@github.com> wrote:
> 
> In docs/design/simple-rolling-update.md:
> 
> > +
> > +Recovery is achieved by issuing the same command again:
> > +
> > +```
> > +kubectl rolling-update rc foo [foo-v2] --image=myimage:v2
> > +```
> > +
> > +Whenever the rolling update command executes, the kubectl client looks for replication controllers called ```foo``` and ```foo-next```, if the exist, an attempt is
> > +made to roll ```foo``` to ```foo-next```.  If ```foo-next``` does not exist, then it is created, and the rollout is a new rollout.  If ```foo``` doesn't exist, then
> > +it is assumed that the rollout is nearly completed, and ```foo-next``` is renamed to ```foo```.  Details of the execution flow are given below.
> > +
> > +
> > +### Aborting a rollout
> > +Abort is assumed to want to reverse a rollout in progress.
> > +
> > +```kubectl rolling-update rc foo [foo-v2] abort```
> switched to a flag.
> 
> I think that --abort is different than --rollback
> 
> abort is more of an "oops, I did the wrong thing, I need to stop and go back"
> 
> rollback is more of an "oh, it ran for N hours but we found a bug" and in that case, it's actually just a new rollout.
> 
> But I'm open to combining them if people feel strongly.
> 
It's a good distinction.  We have something similar - cancel deployment which leaves everything as is, and you can then delete the new rc and pods if you want.  
> —
> Reply to this email directly or view it on GitHub.
>